# 데이터셋 이해

In [1]:
import pandas as pd
train= pd.read_csv('korean-hate-speech/labeled/train.tsv' ,sep='\t')
dev= pd.read_csv('korean-hate-speech/labeled/dev.tsv' ,sep='\t')

In [26]:
train

comments  contain_gender_bias  \
0     (현재 호텔주인 심정) 아18 난 마른하늘에 날벼락맞고 호텔망하게생겼는데 누군 계속...                False   
1     ....한국적인 미인의 대표적인 분...너무나 곱고아름다운모습...그모습뒤의 슬픔을...                False   
2     ...못된 넘들...남의 고통을 즐겼던 넘들..이젠 마땅한 처벌을 받아야지..,그래...                False   
3                    1,2화 어설펐는데 3,4화 지나서부터는 갈수록 너무 재밌던데                False   
4     1. 사람 얼굴 손톱으로 긁은것은 인격살해이고2. 동영상이 몰카냐? 메걸리안들 생각...                 True   
...                                                 ...                  ...   
7891                                      힘내세요~ 응원합니다!!                False   
7892                             힘내세요~~삼가 고인의 명복을 빕니다..                False   
7893                              힘내세용 ^^ 항상 응원합니닷 ^^ !                False   
7894  힘내소...연기로 답해요.나도 53살 인데 이런일 저런일 다 있더라구요.인격을 믿습...                False   
7895                                 힘들면 관뒀어야지 그게 현명한거다                False   

        bias  hate  
0     others  hate  
1       none  none  
2       none  hate  
3       none  none  
4     gender  hate  
...      ...   ...  
7891    none  none  
7892    none  none  
7893    none  none  
7894    none  none  
7895    none  none  

[7896 rows x 4 columns]

## 종속변수 편향 검사
1. social bias : gender, others none
2. hate speech : none, offensive, hate

In [15]:
train['hate'].value_counts()

none         3486
offensive    2499
hate         1911
Name: hate, dtype: int64

In [14]:
train['bias'].value_counts()

none      5148
others    1516
gender    1232
Name: bias, dtype: int64

In [24]:
train[train['bias'].apply(lambda x: x=='gender')].contain_gender_bias.value_counts()

True    1232
Name: contain_gender_bias, dtype: int64

- gender에 대한 bias가 있는 1232명은 contain_gender_bias 값이 True로 할당되어 있음
- 즉 contain_gender_bias column은 삭제해도됨

In [29]:
train[train['bias'].apply(lambda x: x=='others')].hate.value_counts()

offensive    763
hate         616
none         137
Name: hate, dtype: int64

- others에 대해 bias가 있는 사람 중 offensive, hate가 아닌 경우가 존재

In [38]:
others_bias = train[train['bias'].apply(lambda x: x=='others')]
others_bias[others_bias['hate'].apply(lambda x: x =='none')]

comments  contain_gender_bias  \
10                    10살 차이가 넘을텐데 부부라고? 무슨 내용인지 긍금하네..                False   
17    17년도 아니고 27년이면 진짜 너무했다 이것이 사랑만으로 가능할수 있는지 도저히 ...                False   
44                                     20년째 열애중? 너무 길다.                False   
69                    38세48세로 잘못적은거아닌가ㅋㅋㅋㅋ할정도로 동안이네요~~~                False   
92                                          52면 날씬한건데..                False   
...                                                 ...                  ...   
7671                 했다에 내손모가지를 걸지!한게 나쁜거냐?!거짓말하는게 나쁘지!                False   
7675                           행복해보이네. 근데 아기가 없는게 안타깝다.                False   
7695                                         헐 젊은 나이에 왠                False   
7701                  헐~ 스토리상 엄청 연하남인거임? 장미희 남편 연배는 아닌데                False   
7720  현빈 김주원때 좋아하고 좋아한적 없는데 나이먹으니 더 깊이있어지고 멋있네.. 연기도...                False   

        bias  hate  
10    others  none  
17    others  none  
44    others  none  
69    others  none  
92    others  none  
...      ...   ...  
7671  others  none  
7675  others  none  
7695  others  none  
7701  others  none  
7720  others  none  

[137 rows x 4 columns]

In [39]:
others_bias[others_bias['hate'].apply(lambda x: x =='offensive')]

comments  contain_gender_bias  \
15    15살차이면 어릴때사고쳤으면 딸벌인데 ㅋㅋ 능력좋네 전현무 여자들이 전현무스타일이 ...                False   
36    200억대 쳐바르고 홍보를 그렇게나했는데 시청률 5.0%가 자랑이라고 올렸냐... ...                False   
46                                     20대다. 연애 많이 해봐라.                False   
47                                            20살차?? ㅉㅉ                False   
49                           24살인데 술을먹어 발랑까졋네 고딩때도 먹엇을듯                False   
...                                                 ...                  ...   
7823                                          화사는 진짜 못생                False   
7833                              확실히 유승준 엠씨몽보다 지능적이긴했어                False   
7835  환하나 지독한 인간 인 것 같네요. 박유천은 피해자이다. 이외에도 더 많은 피해자가...                False   
7837                환희는 선플로가고 준희는 악플로가는데 생긴걸로 억울한면이 많을듯                False   
7839                                   황교익이 이 기사를 싫어합니다                False   

        bias       hate  
15    others  offensive  
36    others  offensive  
46    others  offensive  
47    others  offensive  
49    others  offensive  
...      ...        ...  
7823  others  offensive  
7833  others  offensive  
7835  others  offensive  
7837  others  offensive  
7839  others  offensive  

[763 rows x 4 columns]

In [40]:
none_bias = train[train['bias'].apply(lambda x: x=='none')]
none_bias[none_bias['hate'].apply(lambda x: x =='offensive')]

comments  contain_gender_bias  \
8                    10년만에 재미를 느끼는 프로였는데왜 니들때문에 폐지를해야되냐                False   
9     10년차방탄팬인데 우리방탄처럼 성공은못하겠지만 일단 방탄의 부하가되고싶다는거니 이름...                False   
13             1400억 뜯기고 얼마나 억울 했을꼬 뒤질만큼 아플만 하지~~~~ㅋㅋㅋㅋ                False   
23                                    1년연예 ㅋㅋ 2년봅니다....                False   
26                                  1보다 재미없어요 진짜 보다 나올뻔                False   
...                                                 ...                  ...   
7840                           황금빛 약빨 떨어지니까 슬금슬금 나오는 군ㅋ                False   
7862                                          흐아아아암 잘잤다                False   
7872                                     흥국아 오래 살았다 아이가                False   
7877                 히든싱어도 끝났다...예전에 비해 가수의 레벨이 너무 떨어짐.                False   
7882                      힘내라 이쁜 혜선아 다음엔 저런인간 만나지 말고^^;                False   

      bias       hate  
8     none  offensive  
9     none  offensive  
13    none  offensive  
23    none  offensive  
26    none  offensive  
...    ...        ...  
7840  none  offensive  
7862  none  offensive  
7872  none  offensive  
7877  none  offensive  
7882  none  offensive  

[1369 rows x 4 columns]

- other bias의 경우 데이터 라벨링이 올바르지 못한 경우가 존재
    - "흥국아 오래 살았따 아이가" : bias - none
    - "	20대다. 연애 많이 해봐라." : bias - others
- 따라서, bias는 gender_bias 가 True/False 로 이진분류로 변환해주고자 함
    - **즉, column을 'bias'가 아닌 'contain_gender_bias'를 이용**

In [42]:
gender_bias = train[train['bias'].apply(lambda x: x=='gender')]
gender_bias[gender_bias['hate'].apply(lambda x: x =='none')]

comments  contain_gender_bias  \
51    25년전이면오달수도 혈기왕성할때인데 그럴수도 있을듯 그리고 이제와서 25년일을 말하...                 True   
452                     거 예비며느리가 빚 청산해주구 시부모한테 이쁨좀 받자 ㅋ                 True   
846                       그래도 남자로서 부럽네...잘생겨서 항상 인기 많은듯                 True   
864                                     그래도 초희 이쁘니까 괜찮아                 True   
949   그분들 몰려오겠네 에휴, 남한테 피해주는것도 아니고 본인이하고싶어서 하는건데 뭐라고...                 True   
...                                                 ...                  ...   
7158                최근 본 사회면 기사중에 유.일.하.게. 여성 댓글 비율이 높네                 True   
7540           한국여자들이 욕하는걸 보니까 송혜교씨 정말 훌륭한 사람이시네요 응원합니다                 True   
7612                                한채아가 아깝다 좋다는 남자 많은데                 True   
7812                    화나요는 다 여자가누른듯, , ,,인정하면 up, , ,                 True   
7814                                화목한 가정 특징 : 마누라가 이쁨                 True   

        bias  hate  
51    gender  none  
452   gender  none  
846   gender  none  
864   gender  none  
949   gender  none  
...      ...   ...  
7158  gender  none  
7540  gender  none  
7612  gender  none  
7812  gender  none  
7814  gender  none  

[76 rows x 4 columns]

In [5]:
texts = train['comments']

In [6]:
import time
from konlpy.tag import Kkma, Twitter, Komoran
pos_taggers = [('kkma', Kkma()), ('twitter', Twitter()), ('Komoran', Komoran())]
results = []
for name, tagger in pos_taggers:
    tokens = []
    process_time = time.time()
    for text in texts:
        tokens.append(tagger.pos(text))
    process_time = time.time() - process_time
    print('tagger name = %10s, %.3f secs' % (name, process_time))
    results.append(tokens)

C:\Users\rlagy\anaconda3\lib\site-packages\konlpy\tag\_okt.py:16: UserWarning: "Twitter" has changed to "Okt" since KoNLPy v0.4.5.
  warn('"Twitter" has changed to "Okt" since KoNLPy v0.4.5.')


tagger name =       kkma, 206.511 secs
tagger name =    twitter, 21.144 secs
tagger name =    Komoran, 4.718 secs
